FR #956: Provider-calculated split cost allocation (#1394)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/allocatedresourcedetails.md

# Allocated Resource Details

> Allocated Resource Details provides information about how resources are allocated when using split cost allocation.

Reviewer: nit: Do we need to introduce the term "split cost allocation" here? I worry that's adding complexity with unnecessary jargon. Can we just say something along the lines of "when usage records are split to support cost allocation requirements."?
```suggestion
Allocated Resource Details provides information about how resources are allocated when usage records are split to support cost allocation requirements.
```

The AllocatedResourceDetails column adheres to the following requirements:

* AllocatedResourceDetails SHOULD be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports [provider-calculated split cost allocation](#provider-calculated-split-cost-allocation).

> * AllocatedResourceDetails MUST be of type String.

Reviewer: Do we plan to have a separate PR to evaluate our use of string vs. object? I just want to make sure this is covered in that.

Reviewer 2: I think that is addressed in PR 1383 specifically here: https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/d14cc9b464cba685996cd6abb0e3cf9d0d8701d2/specification/attributes/json_object_format.md?plain=1#L21

Let me know if it does not.

Reviewer 2: Addressed in TF2 and agreed no action need be taken on this PR.

* AllocatedResourceDetails MUST conform to [StringHandling](#stringhandling) requirements.
* AllocatedResourceDetails MUST conform to [ObjectFormat](#objectformat) requirements.
  * AllocatedResourceDetails MUST conform to ObjectFormat schema requirements below.
* AllocatedResourceDetails nullability is defined as follows:
  * AllocatedResourceDetails MUST be null when a charge is not related to a provider-calculated split cost allocation.

>   * AllocatedResourceDetails SHOULD NOT be null when a charge is related to a provider-calculated split cost allocation.

Reviewer 3: As mentioned in [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1394/files#r2318478128), I assume the following applies to all Allocated-* columns except AllocatedMethodId:

- [AllocatedColumn] MUST be null when a *charge* represents the unallocated portion of the origin *charge* after split cost allocation.

```suggestion
  * AllocatedResourceDetails MUST be null when a [*charge*](#glossary:charge) is not related to a provider-calculated split cost allocation.
  * AllocatedResourceDetails MUST be null when a *charge* represents the unallocated portion of the origin *charge* after split cost allocation.
  * AllocatedResourceDetails SHOULD NOT be null when a *charge* represents the allocated portion of the origin *charge* assigned to a specific sub-resource.
```

Author: Linking related threads:
https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1394#discussion_r2319517990

Reviewer 3: Discussed in TF-2 and in an ad-hoc meeting on Wed, Sep 3. AllocatedResourceDetails, especially AllocatedRatio, provides key information in the case of the unallocated portion of the origin charge after split cost allocation. Following that discussion, I even support specifying that:

> AllocatedMethodDetails **MUST NOT** be null when a charge is related to a provider-calculated split cost allocation.

Anyway, resolving this.

## Object Schema

Allocated Resource Details consists of a valid JSON object which contains an array of key-value objects describing the one or more factors that determined the split cost allocation. Each object consists of FOCUS-defined keys but can be extended to provide additional details about the allocation.

> If AllocatedResourceDetails is not null, the ObjectFormat for AllocatedResourceDetails adheres to the following requirements:

Reviewer: I'm not looking at existing examples of this, but shouldn't we show this as a "When..." bullet up above? 

I'm not holistically opposed to this being here, but I would prefer keeping all requirements in a single place. It's very easy for someone to stop reading after the first set and miss this section entirely. I almost did 😣

Reviewer 2: I believe this is the first time we're defining a schema anywhere, which is why I think it's standing out as odd. This is probably a more wide-ranging discussion related to the ObjectFormat PR as well as this one.

Reviewer 2: Discussed in TF2 and addressed here: https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/3d069a41cdbfc99a3934a0447ca4385db6edbc54/specification/datasets/cost_and_usage/columns/allocatedmethoddetails.md?plain=1#L33

> * ObjectFormat for AllocatedResourceDetails MUST be a valid JSON object.

Reviewer: Isn't this a requirement for ObjectFormat? It shouldn't need to be stated here.

Reviewer 2: >Object Format columns MUST contain a serialized JSON string, consistent with the [ECMA 404](https://www.ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf) definition of an object.

I believe you're correct. I'm okay with dropping it since there is a link to the ObjectFormat definition.

Reviewer 4: Agree.  I think I have a separate comment on this point.  Please cut and then resolve.

Reviewer 2: Deleted.

* AllocatedResourceDetails MUST have a top-level key "Elements" which contains an array.
* Each item in "Elements" MUST be an object.
  * Objects inside "Elements" MUST conform to [KeyValueFormat](#key-valueformat) requirements.

>   * Objects inside "Elements" MUST contain key-value pairs (allocation properties).

Reviewer: nit: Is this line helpful? It seems duplicative to the previous line.

Reviewer 2: Agreed and deleted.

    * FOCUS-defined allocation properties adhere to the following additional requirements:
      * Allocation property key MUST match the spelling and casing specified for the FOCUS-defined property.
      * Allocation property value MUST be of the type specified for that property.

>       * Allocation property MUST adhere to additional normative requirements specific to that property.

Reviewer: nit: Should this be plural? It's not flowing well for me.
```suggestion
      * Allocation properties MUST adhere to additional normative requirements specific to that property.
```

Reviewer 2: Yep, updated locally will push soon.

    * Allocation property keys MUST begin with the string "x_" unless it is a FOCUS-defined allocation property.

> * AllocatedResourceDetails root object MAY contain additional items, in addition to "Elements".

Reviewer: nit: Is there a scenario for this? Not opposed to it, just curious. It feels like we're making the object just a little bit more complicated than it needs to be by using an object with an array of objects rather than a simple array of objects.

Reviewer 2: Discussed in TF2 and decided we don't want to shut the door on this because we don't _know_ it's something that a provider wouldn't want now or in the future.

### FOCUS-Defined Allocation Properties

The following keys are used for allocation properties to facilitate querying data across allocations and across providers. Focus-defined keys will appear in the list below and Provider-defined keys will be prefixed with "x_" to make them easy to identify as well as prevent collisions.

FOCUS-defined allocation properties adhere to the following requirements:

> * "AllocatedRatio" MUST be included inside each "Elements" object.

Reviewer: @kk09v In your mind, does this apply to Cosmos DB? I'm just curious what your expectation is. It's not precisely "split cost allocation", but I know you've used it as a reference for the general space. I just want to make sure I'm looking at this holistically.

Reviewer 2: Will be addressed in the future: https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/1479

>   * Values for "AllocatedRatio" MUST be a decimal value compatible with [NumericFormat](#numericformat) representing the allocated charge's percentage of the origin charge.

Reviewer: nit: I would argue this is likely worth adding a PercentageFormat to be explicit and ensure consistency. Not going to hold the line for it, but calling it out. (See my comment about `1.0` vs. `100`.)

Reviewer 2: Addressed in previous comment.

>   * Values for "AllocatedRatio" across all allocated records related to a single origin record MUST sum up to 1 (100%).

Reviewer 3: > * Values for "AllocatedRatio" across all allocated records related to a single origin record MUST sum up to 1 (100%).

**QUESTION:** Does this sum up to 1 (100%) **only if there is no charge record representing the unallocated portion**, or am I misunderstanding the AllocatedRatio definition?

> AllocatedRatio : Percentage of overall cost derived from corresponding method and metric.

---
Notes:

- Sample data for different scenarios (including those with unallocated charges) is available in [this spreadsheet](https://docs.google.com/spreadsheets/d/1Q14IC3Z_1ch8ZXGcw0VpgzJQ6qGjGYIRDMWOp0u6lv0/edit?gid=72403288#gid=72403288) (sheet [Split Lines Scenario Examples](https://docs.google.com/spreadsheets/d/1Q14IC3Z_1ch8ZXGcw0VpgzJQ6qGjGYIRDMWOp0u6lv0/edit?gid=72403288#gid=72403288)). 
  - These sample numbers were set up according to my understanding of the AllocatedRatio definition; if I misunderstood anything, @kk09v @qquigley **please advise how they should be adjusted**. 

- In addition, the [Split Metrics Calculation for Composite Scenario](https://docs.google.com/spreadsheets/d/1Q14IC3Z_1ch8ZXGcw0VpgzJQ6qGjGYIRDMWOp0u6lv0/edit?gid=1047487568#gid=1047487568) sheet provides a detailed breakdown of the calculations for the metrics we are splitting, illustrated with a scenario using composite usage metrics (Requests and GB). It shows how these numbers were derived.

Author: We can discuss this today to decide whether to update the PR or the spreadsheet, but in my mind Allocated Charge $ = Origin Charge $ * AllocationRatio. I don't care whether it's allocated usage or unallocated usage. 

In general, I think AllocatedResourceId must be null to indicate that it's unallocated cost/usage, but other metadata about the unallocated charges is still useful.

Reviewer 3: Discussed in TF-2 on Wed, Sep 3. We might tweak the Allocated Ratio column definition, and maybe even normative requirements of all 3 properties for clarity, but that’s a [separate conversation](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1394/files#r2337686735
). Resolving this.

>   * Values for "AllocatedRatio" across all allocated records related to a single origin record MUST sum up to 1 (100%).

Reviewer: This implies that numbers are fractions. Is that stated anywhere? Or is this that requirement? I only ask because I've seen many systems where percentages are represented as whole numbers (or at least 0-100, as opposed to 0.0 to 1.0).

Reviewer 2: I see this as that requirement. Should it be more clear?

Reviewer 2: Addressed in the requirements. Re-open if still unclear.

> * "UsageUnit" MUST be included inside an "Elements" object if "UsageQuantity" allocation property is included in that "Elements" object, otherwise "UsageUnit" MAY be included in each "Elements" object.

Reviewer: Why not `ConsumedUnit` for consistency? Ideally we should be consistent if they're the same. If they're not the same, then we should probably come up with a different name because this will lead to confusion since people think of "ConsumedUnit" as "UsageUnit" already.

Reviewer 2: UsageUnit represents the allocation of the origin record's ConsumedUnit.

  * Values for "UsageUnit" SHOULD capture the unit or component of provider's documented [AllocationMethod](#allocationmethod) that was used to determine the "AllocatedRatio" value.

>   * Values for "UsageUnit" are RECOMMENDED to use [UnitFormat](#unitformat).

Reviewer 3: **A few questions:**

- Why is it specified as SHOULD rather than MUST in this requirement: "...Values for 'UsageUnit' SHOULD capture the unit or component..."?

- Do you agree that for all three properties, MUST NOT be null applies when a charge represents the allocated portion of the origin charge?

```suggestion
  * Values for "UsageUnit" MUST capture the unit or component of provider's documented [AllocationMethod](#allocationmethod) that was used to determine the "AllocatedRatio" value.
  * Values for "UsageUnit" SHOULD conform to [UnitFormat](#unitformat) requirements.
```

---

BTW, for now, my comments on FOCUS-Defined Allocation Properties are mainly focused on the content. 

Once the structure of the definitions/descriptions for the JSON columns AllocatedResourceDetails and Contract Applied (see PR #1350) has been aligned, we can analyze them from the perspective of the Column Normative Guidelines (assuming that, ultimately, these requirements should follow the guidelines as closely as possible).

> * "UsageQuantity" MAY be included inside an "Elements" object when that "Elements" object contains a "UsageUnit" allocation property.

Reviewer: Same: Why not `ConsumedQuantity` for consistency?

Reviewer 2: Addressed above.

  * Values for "UsageQuantity" MUST be compatible with NumericFormat.
  * Values for "UsageQuantity" SHOULD capture the quantity or volume of the "UsageUnit" measured by the provider that was used to determine the "AllocatedRatio" value.

> | Key | ValueType | Required | Description |

Reviewer 4: I've proposed a reorg of this section per a branch I've created [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/956-allocated-resource-details-revision/specification/datasets/cost_and_usage/columns/allocatedresourcedetails.md).

I haven't issued a PR, but I can (or @kk09v can) if it's helpful!

Author: I took a look and generally like it.

Reviewer 4: I believe we've circled the wagons on this one sufficiently.  Resolving!

| ----- | ---- | ---------- | ----------- |
| Elements | Array | TRUE | The parent array containing one or more objects which communicate information about how an allocated record was calculated. |
| AllocatedRatio | Numeric | TRUE | Percentage of overall cost derived from corresponding method and metric. |
| UsageUnit | [String](#stringhandling) | Conditional | Unit being measured used to calculate allocation. |
| UsageQuantity | Numeric | FALSE | Volume of UsageUnit consumed or used. |

> ### Example

Reviewer 4: Would be good to have another example that shows when some subset of the object entries aren't included by the provider (e.g., UsageUnit).

Author: Acknowledged. I want to update the example with info as well and re-re-validate the JSON / JTD

Reviewer 4: Just to confirm, this thread still needs to be addressed.

Reviewer 2: Addressed: https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/3d069a41cdbfc99a3934a0447ca4385db6edbc54/specification/datasets/cost_and_usage/columns/allocatedmethoddetails.md?plain=1#L174

# Allocated Resource Details

Allocated Resource Details provides information critical in understanding how resources are allocated when using split cost allocation. This information includes: allocation method name, metric used to calculate cost, the usage value/quantity, and the ratio of the cost derived from the method.

> | Parent | Key | Type | Extensible | Description | Example |

Author: Trying to understand the overall shape here. Method has a parent of Null, has [key-value format](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/attributes/key_value_format.md), and also has children (Metric has parent Method)

This would be what it would look like if all columns had a parent of NULL
{"Method":"MethodName", "Metric":"CPU", "UsageValue":0.5, "RatioValue":0.05}

I feel like we're missing an object to hold the the key-value pairs currently pointing to Method as a parent. Does this feel right?
{
"Method":"MethodName", 
"Calculation": [
    {
    "Metric":"CPU", 
    "UsageValue":0.5, 
    "RatioValue":0.05
    },
    {
    "Metric":"Memory", 
    "UsageValue":4, 
    "RatioValue":0.1
    }
]
}

If so, this would be more like:
| null | Method | Key-Value | FALSE | Name of allocation method used by the provider. | Method: methodname |
| null | Calculation | Object | TRUE | Object which contains the metrics used to calculate provider-calculated split cost allocation. | Calculation: [...] |
Then change parent from Method to Calculation.

I do think we need to add a complete object to this PR as an example.

Author: GitHub squashed my beautiful JSON formatting :(

Author: Our [numeric format](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/attributes/numeric_format.md) allows for multiple precisions, but I believe I have to declare it explicitly for a valid JTD, so I went with float64. I'm not an expert in JTD, but I believe this matches the sample I put above.
`{
  "properties": {
    "Method": { "type": "string" },
    "Calculation": {
      "elements": {
        "properties": {
          "Metric": { "type": "string" },
          "UsageValue": { "type": "float64" },
          "RatioValue": { "type": "float64" }
        }
      }
    }
  }
}`

Author: It squashed my formatting again, this time as code. 
Fool me once, shame on you. Fool me twice, shame on me for even trying a second time.

Reviewer 2: I'm following and agree. I'll try to get these edits in before TF2 today.

Reviewer 2: How do we feel about the latest commit? Can I resolve this conversation?

| ------ | --- | ---- | ---------- | ----------- | ------- |

> | null | Method | Key-Value | TRUE | Allocation method used by the provider. | method: methodname |

Author: ```suggestion
| null | Method | Key-Value | TRUE | Allocation method used by the provider. | Method: methodname |
```
Aligning the example with the defined key

| Method | Metric | Key-Value | FALSE | Metric used to calculate allocation. | Metric: CPU |
| Method | UsageValue | Key-Value | FALSE | Measured units. | UsageValue: 0.5 |

> | Method | RatioValue | Key-Value | FALSE | Ratio of overall cost derived from method. | RatioValue: .05 |

Author: ```suggestion
| Method | RatioValue | Key-Value | FALSE | Ratio of overall cost derived from method. | RatioValue: 0.05 |
```
Adding leading 0 for ease of reading

# Allocated Resource Details

Allocated Resource Details provides information critical in understanding how resources are allocated when using split cost allocation. This information includes: allocation method name, metric used to calculate cost, the usage value/quantity, and the ratio of the cost derived from the method.

| Parent | Key | Type | Extensible | Description | Example |
| ------ | --- | ---- | ---------- | ----------- | ------- |
| null | Method | Key-Value | TRUE | Allocation method used by the provider. | Method: methodname |
| Method | Calculation | Object | TRUE | Object which contains the metrics used to calculate provider-calculated split cost allocation. | n/a |
| Calculation | Metric | Key-Value | FALSE | Metric used to calculate allocation. | Metric: CPU |
| Calculation | UsageValue | Key-Value | FALSE | Measured units. | UsageValue: 0.5 |

> | Calculation | RatioValue | Key-Value | FALSE | Ratio of overall cost derived from method. | RatioValue: 0.05 |

Reviewer 3: ### Naming Suggestions

Consider renaming:

- **Metric → Meter** — following the pattern of SKU Meter (“the functionality being metered or measured by a particular SKU in a charge”).
- **UsageValue → Quantity** (maybe with a prefix), following the Pricing and Consumed Quantity

### Consider Adding:

- **Unit** (maybe with a prefix), following the Pricing and Consumed Units

### Question

- Does `RatioValue` equal `UsageValue` (i.e. `Quantity`) × `Weight` — or did I misunderstand?  
  - If yes, I’d prefer **Weight**, since that’s the actual ingredient — but as long as the info is available, it’s fine.

### Sample data

Just flagging that we have sample data in the spreadsheet:

- See sheet [Split Lines Scenario Examples](https://docs.google.com/spreadsheets/d/1Q14IC3Z_1ch8ZXGcw0VpgzJQ6qGjGYIRDMWOp0u6lv0/edit?gid=72403288#gid=72403288)  
- And sheet [Weighted Resource Usage Allocation](https://docs.google.com/spreadsheets/d/1Q14IC3Z_1ch8ZXGcw0VpgzJQ6qGjGYIRDMWOp0u6lv0/edit?gid=1047487568#gid=1047487568) shows how the numbers in the `Allocated` rows were derived — focusing on the **Multi–Allocation Meter** scenario.

---

*Sharing a first attempt to describe the allocation method in words below:*

# Allocation Model

Allocation Model describes how to allocate a usage-based **metric** (e.g., cost, quantity) across multiple subresources.

## Definitions

- **`n`**: number of subresources  
- **`m`**: number of **Allocation Meters** used to evaluate each Allocation Meter (e.g., Cost, Quantity)  

For each **resource `i`**, there is an **Allocation Meter**, which includes:

- **`q_{i,j}`**: quantity of Allocation Meter `j` for subresource `i`  
- **`w_{i,j}`**: weight of Allocation Meter `j` for subresource `i`  
- **`u_{i,j}`**: unit of Allocation Meter `j` for subresource `i` *(optional, for reporting)*  

**Weighted usage** for subresource `i` is calculated as:

$$
A_i = \sum_{j=1}^{m} q_{i,j} \cdot w_{i,j}
$$

Total weighted usage across all subresources:

$$
A_{\text{total}} = \sum_{i=1}^{n} A_i
$$

Other parameters:

- **`M_used`**: portion of the metric (e.g., cost, quantity) that is usage-based
- **`M_idle`**: portion of the metric (e.g., cost, quantity) that is idle (not distributed to subresources)  
- **`M_i`**: share of the metric (e.g., cost, quantity) allocated to subresource `i`

## Allocation Formula (Usage-Based Only)

If the **idle portion remains with the parent** and is **not distributed**, the share for subresource `i` is:

$$
M_i = \frac{A_i}{A_{\text{total}}} \cdot M_{\text{used}}
$$

## Allocation Formula (Including Idle Portion)

If the **idle portion is evenly distributed** across all Allocation Meters, the formula becomes:

$$
 M_i = \left( \frac{A_i}{A_{\text{total}}} \cdot  M_{\text{used}} \right) + \left( \frac{ M_{\text{idle}}}{n} \right)
$$

## Notes

- Allocation Meters are inputs used to calculate the weighted usage for each resource.  
- Allocation Meter consists of metric quantities, weights, and optionally units.  
- The model supports any number of Allocation Meter per resource.

Author: Method -> AllocationMethodId
UsageQuantity
UsageUnit
AllocationRatio?

Reviewer 3: In the end TF-2 agreed on these terms:

- AllocaatedMethodId
- AllocatedMethodDetails
	- UsageQuantity
	- UsageUnit
	- AllocatedRatio
- Allocated*

Resolving this conversation

The AllocatedResourceDetails column adheres to the following requirements:

* AllocatedResourceDetails SHOULD be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports [provider-calculated split cost allocation](#provider-calculated-split-cost-allocation).
* AllocatedResourceDetails MUST be of type String.
* AllocatedResourceDetails MUST conform to [StringHandling](#stringhandling) requirements.
* AllocatedResourceDetails nullability is defined as follows:
  * AllocatedResourceDetails MUST be null when a charge is not related to a provider-calculated split cost allocation.
  * AllocatedResourceDetails SHOULD NOT be null when a charge is related to a provider-calculated split cost allocation.
* AllocatedResourceDetails maximum length SHOULD be provided in the corresponding FOCUS Metadata Schema.

## Column ID

AllocatedResourceDetails

## Display Name

Allocated Resource Details

## Description

> Self-contained summary of the allocated cost's purpose and price.

Reviewer 3: By convention, the description typically repeats the first sentence from the corresponding introductory section.

Author: Done

  * Objects inside the array MUST conform to [KeyValueFormat](#key-valueformat) requirements.
* Property keys MUST begin with the string "x_" unless it is a FOCUS-defined property.
* FOCUS-defined Allocation properties adhere to the following additional requirements:
  * Property key MUST match the spelling and casing specified for the FOCUS-defined property.
  * Property value MUST be of the type specified for that property.

### FOCUS-Defined Properties

The following keys should be used when applicable to facilitate querying data across allocations and across providers. Focus-defined keys will appear in the list below and Provider-defined keys will be prefixed with "x_" to make them easy to identify as well as prevent collisions.

| Key | ValueType | Required | Description |
| ----- | ---- | ---------- | ----------- |
| AllocatedRatio | [Numeric](#numericformat) | TRUE | Percentage of overall cost derived from corresponding method and metric. |
| UsageUnit | [String](#stringhandling) | FALSE | Unit being measured used to calculate allocation. |
| UsageQuantity | Numeric | FALSE | Volume of UsageUnit consumed or used. |

### Example

```json

> [

Reviewer 4: Examples need to match sample data format (i.e., object of array of objects).

  {
    "AllocatedRatio": 0.05,
    "UsageUnit": "CPU",

>     "UsageValue": 0.5

Reviewer 4: This needs to be Quantity instead of value.

  },
  {
    "AllocatedRatio": 0.1,
    "UsageUnit": "Memory",

>     "UsageValue": 4

Reviewer 4: This needs to be Quantity instead of Value.

  }
]
```

### JSON Type Definition

```json
{
  "elements": {
    "properties": {
      "AllocatedRatio": { "type": "float64" }
    },
    "optionalProperties": {
      "UsageUnit": { "type": "string" },

>       "UsageValue": { "type": "float64" }

Reviewer 4: This needs to be Quantity instead of Value.
